src/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -373,14 +419,22 @@ private void buildSortedReadEndLists(final boolean useBarcodes) {
                 pgIdsSeen.add(rec.getStringAttribute(SAMTag.PG.name()));
             }
 
+            // Of working in query-sorted, need to keep index of first record with any given query-name.
+            if(assumedSortOrder == SAMFileHeader.SortOrder.queryname && !rec.getReadName().equals(duplicateQueryName)) {
+                duplicateQueryName  = rec.getReadName();
+                duplicateIndex      = index;
+            }
+
             if (rec.getReadUnmappedFlag()) {
-                if (rec.getReferenceIndex() == -1) {
-                    // When we hit the unmapped reads with no coordinate, no reason to continue.
+                if (rec.getReferenceIndex() == -1 && assumedSortOrder == SAMFileHeader.SortOrder.coordinate) {
+                    // When we hit the unmapped reads with no coordinate, no reason to continue (only in coordinate sort).
                     break;
                 }
                 // If this read is unmapped but sorted with the mapped reads, just skip it.
+
             } else if (!rec.isSecondaryOrSupplementary()) {
-                final ReadEndsForMarkDuplicates fragmentEnd = buildReadEnds(header, index, rec, useBarcodes);
+                final long indexForRead = assumedSortOrder == SAMFileHeader.SortOrder.coordinate ? index : duplicateIndex;
```

Comment:
Would it be clearer to say:
final long indexForRead = assumedSortOrder == SAMFileHeader.SortOrder.queryname ? duplicateIndex : index;
so that it matches up better with the if on 423?

since we only have the two options, they're equivalent, so this is just a readability thing